File: leetcode/problems/problem-specific/49.group-anagrams/49.group-anagrams.go
package main

func groupAnagrams(strs []string) [][]string {
	if len(strs) <= 0 {
		return [][]string{}
	}

	res, mappings := [][]string{}, map[string][]string{}
	for _, v := range strs {
		characterMappings := make([]int, 26)
		for _, v1 := range v {
			characterMappings[v1-'a']++
		}

		keyString := ""
		for _, v1 := range characterMappings {
			keyString += string(v1)
		}

		mappings[keyString] = append(mappings[keyString], v)
	}

	for _, v := range mappings {
		res = append(res, v)
	}

	return res
}

File: leetcode/problems/problem-specific/49.group-anagrams/49.group-anagrams.py
#
# @lc app=leetcode id=49 lang=python3
#
# [49] Group Anagrams
#

# @lc code=start

# * Hash Table Solution | O(mn) Time | O(n) Space
# * m -> The length of strs array | n -> The length of each string


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        res = collections.defaultdict(list)
        for s in strs:
            counts = [0] * 26
            for char in s:
                counts[ord(char) - ord("a")] += 1
            res[tuple(counts)].append(s)

        return res.values()


# @lc code=end
